src/gui/plugins/environmental_data_loader/EnvironmentalDataLoader.cc

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "EnvironmentalDataLoader.hh"
+
+#include <gz/gui/Application.hh>
+#include <gz/gui/MainWindow.hh>
+#include <gz/sim/components/EnvironmentalData.hh>
+#include <gz/sim/Util.hh>
+
+#include <gz/plugin/Register.hh>
+
+#include <atomic>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include <gz/common/CSVStreams.hh>
+#include <gz/common/DataFrame.hh>
+
+using namespace gz;
+using namespace sim;
+
+namespace gz
+{
+namespace sim
+{
+inline namespace GZ_SIM_VERSION_NAMESPACE
+{
+/// \brief Private data class for EnvironmentalDataLoader
+class EnvironmentalDataLoaderPrivate
+{
+  public: QString dataPath;
+
+  public: QStringList dimensionList;
+
+  public: int timeIndex{-1};
+
+  public: int xIndex{-1};
+
+  public: int yIndex{-1};
+
+  public: int zIndex{-1};
+
+  public: std::mutex mutex;
+
+  public: std::atomic<bool> needsUpdate{false};
+};
+}
+}
+}
+
+EnvironmentalDataLoader::EnvironmentalDataLoader()
+  : GuiSystem(), dataPtr(new EnvironmentalDataLoaderPrivate)
+{
+  gui::App()->Engine()->rootContext()->setContextProperty(
+      "EnvironmentalDataLoader", this);
+}
+
+EnvironmentalDataLoader::~EnvironmentalDataLoader()
+{
+}
+
+void EnvironmentalDataLoader::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Environmental Data Loader";
+
+  gui::App()->findChild<gui::MainWindow *>()->installEventFilter(this);
+}
+
+void EnvironmentalDataLoader::Update(const UpdateInfo &,
+                                     EntityComponentManager &_ecm)
+{
+  if (this->dataPtr->needsUpdate)
+  {
+    std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
```

Comment:
Should we move this lock before the `if` above?

Reply:
@caguero that's the point of using an atomic boolean: avoid unnecessary mutex locking on every simulation step.

Reply:
Ah, I didn't realize `needsLoad` was atomic. Then you can move the `lock` after setting `needsLoad` to false, right?

Reply:
@caguero Yes and no. No in the general case, as you could potentially race with another thread to reach the lock. Yes in this particular case as we know there won't more than one thread exercising this execution path (IIRC there's just one thread stepping the simulation). 

I went for the generally correct case out of habit. We can move it out of the lock if you want.